firebase-firestore/src/main/java/com/google/firebase/firestore/remote/BloomFilter.java

```diff
@@ -0,0 +1,134 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.remote;
+
+import androidx.annotation.NonNull;
+import com.google.firebase.firestore.util.Logger;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+
+public class BloomFilter {
+  private static final String TAG = "BloomFilter";
+
+  private final int size;
+  private final byte[] bitmap;
+  private final int hashCount;
+
+  public BloomFilter(@NonNull byte[] bitmap, @NonNull int padding, @NonNull int hashCount) {
+    if (padding < 0 || padding >= 8) {
+      throw new IllegalArgumentException("Invalid padding: " + padding);
+    }
+
+    if (bitmap.length > 0) {
+      // Only empty bloom filter can have 0 hash count.
+      if (hashCount <= 0) {
+        throw new IllegalArgumentException("Invalid hash count: " + hashCount);
+      }
+    } else {
+      if (hashCount < 0) {
+        throw new IllegalArgumentException("Invalid hash count: " + hashCount);
+      }
+
+      // Empty bloom filter should have 0 padding.
+      if (padding != 0) {
+        throw new IllegalArgumentException("Invalid padding when bitmap length is 0: " + padding);
+      }
+    }
+    this.bitmap = bitmap;
+    this.hashCount = hashCount;
+    this.size = bitmap.length * 8 - padding;
+  }
+
+  @NonNull
+  public int getSize() {
+    return this.size;
+  }
+
+  public boolean isEmpty() {
+    return this.size == 0;
+  }
+
+  public boolean mightContain(@NonNull String value) {
+    // Empty bitmap or empty value should always return false on membership check.
+    if (this.isEmpty() || value.isEmpty()) {
+      return false;
+    }
+
+    byte[] md5HashedValue = this.MD5Hash(value);
+    if (md5HashedValue == null || md5HashedValue.length != 16) {
+      return false;
+    }
+
+    long hash1 = this.getLongLittleEndian(md5HashedValue, 0);
+    long hash2 = this.getLongLittleEndian(md5HashedValue, 8);
+
+    for (int i = 0; i < this.hashCount; i++) {
+      int index = this.getBitIndex(hash1, hash2, i);
+      if (!this.isBitSet(index)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public static byte[] MD5Hash(String value) {
+    try {
+      MessageDigest digest = MessageDigest.getInstance("MD5");
+      digest.update(value.getBytes());
+      return digest.digest();
+    } catch (NoSuchAlgorithmException e) {
+      Logger.warn(TAG, "Could not create hashing algorithm: MD5.", e);
+      return null;
```

Comment:
Rather than returning null, re-throw the exception in a RuntimeException. `MessageDigest.getInstance("MD5")` should _never_ throw because Android always supports "MD5": https://developer.android.com/reference/java/security/MessageDigest. You should just be able to write `throw new RuntimeException(e);`.